### PR TITLE
Small speed improvement to Stitch1D

### PR DESCRIPTION
**Pull request by robertapplin**
### Description of work
This PR makes a small change which results in a very slight speed improvement for the `Stitch1D` algorithm. The `hasNonzeroErrors` function was very slow for what it was doing.

The yellow scribble square indicates the amount of time that has been shaved off the execution of this algorithm (approximately 1/10th the original runtime).
image

Its a very small improvement (0.05 seconds), but this algorithm is run many times when reducing hundreds of files.

### To test:
Code review and passing tests should be enough.

*This does not require release notes* because **speed improvement is so minimal it probably won't be noticed on its own**

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Algorithms/src/Stitch1D.cpp [hunk 1]
```diff
@@ -423,26 +420,14 @@ boost::tuple<int, int> Stitch1D::findStartEndIndexes(double startOverlap, double
...
-    if (!hasNonZeroErrors) // Keep checking
-    {
-      const auto &e = ws->e(i);
-      auto it = std::find_if(e.begin(), e.end(), isNonzero);
-      if (it != e.end()) {
-        PARALLEL_CRITICAL(has_non_zero) {
-          hasNonZeroErrors = true; // Set flag. Should not need to check any more spectra.
-        }
-      }
+  for (auto i = 0u; i < ws->getNumberHistograms(); ++i) {
```
Comment (GuiMacielPereira) [inline Framework/Algorithms/src/Stitch1D.cpp:423]: I really like the detail of using unsigned int in the for loop, I'll start doing that myself from now on :)

**Review by GuiMacielPereira — DISMISSED** [review]
The code is clean and much easier to understand than before. The changes are covered by two unit tests that test the function on a workspace with single spectra and a workspace with multiple spectra. 

The function checks that the workspace has ANY non-zero value, so it could contain rows of zeros (which is a bit surprising since the function is used as a check on whether a weighted mean is run which should not work with zeros). Regardless, this is all covered already by the unit tests so all is well.

Also added `const` at several places which is good.

File: Framework/Algorithms/src/Stitch1D.cpp [hunk 2]
```diff
@@ -423,26 +420,14 @@ boost::tuple<int, int> Stitch1D::findStartEndIndexes(double startOverlap, double
...
+    const auto &e = ws->e(i);
+    // Its faster to sum and then check the result is non-zero than to check each element is non-zero individually
```
Comment (thomashampson) [inline Framework/Algorithms/src/Stitch1D.cpp:425]:
```suggestion
    // It's faster to sum and then check the result is non-zero than to check each element is non-zero individually
```

**Review by thomashampson — APPROVED** [review]
As Gui says, I think this is a neat improvement. Re-approving.

**Pull request merged**